### ISSUE TITLE
New Exchange: EtherDelta

### ISSUE THREAD

**Issue by fredfortier** [ccxt/ccxt#715]
Hi! I noticed that EtherDelta is not in the list. Its API looks fairly standard so it seems like it should be eligible: https://github.com/etherdelta/etherdelta.github.io/blob/master/docs/API.md.

Do you have any objection to including it? If not, I'll create a pull request and add it.

**Comment by kroitor** [comment]
@fredfortier absolutely no objections )) In fact, we had multiple requests for EtherDelta, so, I guess, the time has come ;) 

> If not, I'll create a pull request and add it.

That would be amazing! Let me know if you have any questions... We appreciate all help and we welcome all contributions and PRs! Feel free to submit it, I'll stand by to assist if needed. Also, this might be a good place to start with: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md

Thank you!

**Comment by kroitor** [comment]
Reference: #404

**Comment by fredfortier** [comment]
This library helped me save time so it is my pleasure to contribute. I'm only interested by the Python implementation, but I can try doing it in JavaScript and transpile it if that's the preferred workflow.

I should should have a PR by the end of the week.

**Comment by fredfortier** [comment]
I started working on this. A key consideration is that EtherDelta uses token addresses instead of a set list of currency pairs (aka markets). That's how traders are able to trade ERC20 tokens right after distribution without a formal listing. Unless a more clever idea comes to mind, a static file containing a symbol / token address mapping should do the job.

**Comment by kroitor** [comment]
Currently, we don't have support for Websocket APIs, but we are working on it.

> A key consideration is that EtherDelta uses token addresses instead of a set list of currency pairs (aka markets).

> There's a REST API but it's flagged as deprecated.

Yep, these two considerations are two main problems with integrating that distributed exchange, this is why I placed this link above: https://github.com/ccxt/ccxt/issues/404. I'd suggest to integrate their REST API with a static file, then switch to their WS API later, when we add support for WS to the library.

**Comment by frosty00** [comment]
I've actually written a bot that trades using etherdelta so I am familiar with the ins and outs of the exchange. 

**A few things that you may want to consider before writing the interface:** 

1. When depositing tokens they must invoke the depositToken function (if they are sent normally they will be lost forever) - so the fetch_deposit_address should issue a warning

2. As of now only has a websocket API documented, but there is another regular API that is no longer documented but still works (note: this may get depreciated in the future but I have been using it for two months with no problem) 

[This is the regular WS API ](https://github.com/etherdelta/bots/tree/master/python) (maybe you could emulate it to look like regular http requests for the user)

The old API that is no longer documented is 'https://api.etherdelta.com/orders/{}/0' where the token address replaces the square brackets. You can get the token addresses from the website via another undocumented API call which is 'https://api.etherdelta.com/returnTicker'

3. The server is very laggy so 50% or more of http requests will fail

4. You would have to write another general ethereum interface to create the transactions and post them to the network. You could use web3 but I would create another module for this.

Guys I am very busy atm, but I would love to work with you (in python) next week and over Christmas on this project.

**Comment by kroitor** [comment]
@frosty00 

> Guys I am very busy atm, but I would love to work with you (in python) next week and over Christmas on this project.

That would be awesome!

**Comment by fredfortier** [comment]
I spoke with the EtherDelta people. They confirmed @frosty00 's comment that the REST api is unavailable. In my opinion, this downside is significant enough to rule it out.

@frosty00 the project which you linked looks great! A python wrapper is what I need as well for my project but I wanted to contribute to CCXT to help its community.

I suggest that we have these options:

1. Modify the Python WS wrapper linked by @frosty00 to directly create a Python implementation.
2. Wait until CCXT finalizes its support of WS interfaces. Then, write the EtherDelta wrapper in JavaScript, and transpile to other implementations. 

@kroitor do you feel like option 1 (Python only implementation) would be valuable? If so, I'm happy give a stab at it (potentially with some help from @frosty00). If not, I'll add the exchange to my project directly. I would be glad to help later when the WS framework is ready.

**Comment by kroitor** [comment]
> @kroitor do you feel like option 1 (Python only implementation) would be valuable? 

I think, it would, however, we have to be careful with this, because each added feature introduces the need to keep everything consistent and backward-compatible. I think we'd rather go with plan 2:

> If not, I'll add the exchange to my project directly.

Because the work on WS has been started already, I'd suggest you to take a look here: https://gist.github.com/kroitor/7dce1d23a10937ab8c07a5451f17ccf2
These are design considerations for supporting WS in the library.
Follow-up: https://github.com/ccxt/ccxt/pull/751
So, if you could catch up with the above and follow those design considerations, that would be awesome. If you'd suggest to add or change something, let us know.

> I would be glad to help later when the WS framework is ready.

We really appreciate your effort! Thanks so much! We will keep you updated on the progress of WS, let us know if you have any questions with the lib in the meantime. I would close this for now, if you don't mind. Feel free to reopen it or just continue there discussion here.

**Comment by frosty00** [comment]
So I have been running a bot [trading there](https://medium.com/@carlorevelli/how-to-make-your-own-trading-bot-83b5c6e35036) for quite a while now. It has recently gotten a lot worse, the API barely responds and the people running the site seem quite dodgy. Also the volume has dropped loads - maybe people are using other exchanges?

Perhaps it would be better to focus on adding better centralised exchanges or even IDEX rather than etherdelta. I am also afraid that the new owners will maliciously send people bad data so that they can steal funds.

I am actively using ccxt to make a trading bot rn, and once that is finished I could help you work on adding more niche decrentralised exchanges to the API, and even possibly adding features like broadcasting eth transactions natively.

**Comment by kroitor** [comment]
Thanks for the feedback, @frosty00 ! Really insightful! We also hope to throw most of our time at this task in nearest future (after stabilizing and integrating the WS). Having the ability to place orders (txs) on-chain without the need in any relay would be fantastic )

**Comment by fredfortier** [comment]
I believe that RadarRelay is benefiting from the EtherDelta mishaps.

**Comment by kroitor** [comment]
@fredfortier yep, it also came across my mind... At least, they should )

**Comment by lfern** [comment]
I would like to contribute in dex exchanges. I think that integrating any dex api won't be easy, but I am still learning about it. I think it should be necesary to develop:

- some kind of wallet module to interact with every blockchain network (to send transactions). 
- a set of protocol modules: 0x, counterparty, ethereum-erc20, ripple, ...
- changes on base/Exchange to make use of this modules, and then dex exchanges classes that inherits from base.
- add websockets support

Anyway I would like to learn about it, so if you have any skeleton or block diagram I would like to contribute to develop in this issue.

Thanks

**Comment by frosty00** [comment]
@Ifern I've created an etherdelta trading bot already, check the post above. What we would need is a way to publish transactions - pherhaps etherscan's pushTx API. Then we would also have to sign the transactions and encode the data to buy and sell on etherscan. This involves signing some attributes of an order object and sending that transaction to etherdelta. 

Getting past orders is very difficult, you have to look at the ethereum event logs and it is all encoded in hexadecimal. If the order has already been filled by the time the transaction confirms the transaction will fail and you will still have to pay gas prices. IMO this is a seperate project all in itself and it kind of beyond the domain of ccxt.

**Comment by valentinbora** [comment]
@kroitor shall we rename this issue to ForkDelta?

Also, would a very basic fetchTickers implementation help?

I'm looking at combining https://api.forkdelta.com/returnTicker with https://forkdelta.github.io/config/main.json to get some basic data.

**Comment by kroitor** [comment]
> @kroitor shall we rename this issue to ForkDelta?

Nope ) Because ED != FD )

**Comment by valentinbora** [comment]
OK I guess I was partially misled by CMC. In that case we need a new issue.

**Comment by frosty00** [comment]
etherdelta is shutdown for multiple years
